### PR TITLE
Connection Pooling w/ invalidation & with-timeout macro

### PR DESCRIPTION
**Pull request by martinklepsch**
This replaces the existing connection pooling implementation with one that is capable of invalidating connections.

Also using `open-company.db.pool/rebuild-pool!` it is now possible to rebuild the pool in the case RethinkDB has been restarted or similar.

### To Test ([trello card](https://trello.com/c/5ebQjwZS/208-db-pool-improvements))

- Run tests
- Review changes in inidividual files
- Review `default-timeout` (300ms)

Addressing the individual points from the [trello card](https://trello.com/c/5ebQjwZS/208-db-pool-improvements):

---
**return other than :ok for start if it fails to start**

A pool can always be started. If the pool fails to create a resource it will try again in a specified interval (default: 5s). If the pool continously fails to create resources any attempts to claim a resource from the pool will throw an exception.

---
**check for validity of connections (bounce rethinkDB to force them to go bad)**

I looked for a way to check if a connection is healthy but couldn't find any. 
If a connection is bad (causes an exception) the `with-pool` macro will invalidate that resource.
To invalidate all connections there is `open-company.db.pool/rebuild-pool!`.

---
**reconnect them**

Handled by the above(?)

---
**block or otherwise handle the pool being out of connections**

The pool will throw exceptions if no resources are available. This can be mitigated by supplying a timeout (default: 0ms)

---
**log much**

- [x] add a bunch of `timbre/trace` statements

---

**DB pool doesn’t work at 50 connections (server) or 60 connections (locally), where is this limit from?**

There are no hard limits from RethinkDB so I'm not sure what's causing this. @belucid What exactly does "doesn't work" mean here?

@martinklepsch this was the limit in core.async based on CPU cores that you discovered.

---
**API server is sometimes unresponsive after a cold start (any requests that access the DB block indefinitely), and are only fixed by restarting the API server**

Testing this stuff on the REPL I often noticed that closing connections would block indefinitely. On other occasions it just worked again. I guess we should look at this again after we added more logging.

Other notes:

- [x] copy over test suite from riemann project(?)

### CODE REVIEW

File: src/open_company/app.clj [hunk 1]
```diff
@@ -46,6 +47,12 @@
     hot-reload-routes))
 
 (defn start [port]
+  (timbre/set-config! c/log-config)
+  ;; See https://stuartsierra.com/2015/05/27/clojure-uncaught-exceptions
+  (Thread/setDefaultUncaughtExceptionHandler
+   (reify Thread$UncaughtExceptionHandler
+     (uncaughtException [_ thread ex]
+       (timbre/error ex "Uncaught exception on" (.getName thread)))))
```
Comment (martinklepsch) [inline src/open_company/app.clj]: Not really in the scope of this PR but probably just fine it leave it in anyways.

**tl;dr:** Exceptions in threads may not always be handled by default. This snippet makes sure that in every thread there is a handler for uncaught exceptions which just logs them.

Comment (belucid) [inline src/open_company/app.clj]: Don't we want them to go to Sentry @martinklepsch ? not just logged

Comment (martinklepsch) [inline src/open_company/app.clj]: Right, that would make sense, will update.

**Comment by belucid** [comment]
@martinklepsch just waiting on the Sentry change, then I'll review.

File: src/open_company/resources/common.clj [hunk 2]
```diff
@@ -143,6 +144,22 @@
...
+(defmacro with-timeout
+  "A basic macro to wrap things in a timeout.
+  Will throw an exception if the operation times out.
+  Note: This is a simplistic approach and piggiebacks on core.asyncs executor-pool.
+  Read this discussion for more info: https://gist.github.com/martinklepsch/0caf92b5e42eefa3a894"
+  [ms & body]
+  `(let [c# (async/thread-call #(do ~@body))]
+     (let [[v# ch#] (async/alts!! [c# (async/timeout ~ms)])]
+       (if-not (= ch# c#)
+         (throw (ex-info "Operation timed out" {}))
```
Comment (belucid) [inline src/open_company/resources/common.clj]: Isn't the idea @martinklepsch to use invalidate here to get that connection dumped from the pool? Seems this will prevent the long term lockups, but we'll still keep getting the bad connection from the pool.

Comment (belucid) [inline src/open_company/resources/common.clj]: Oh... maybe the exception causes the pool to take this connection out of the pool?

**Comment by belucid** [comment]
Looks good @martinklepsch
